crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -33,11 +34,105 @@ use sp_objects::ObjectsApi;
 use std::sync::Arc;
 use subspace_archiving::archiver::{Archiver, NewArchivedSegment};
 use subspace_core_primitives::crypto::kzg::Kzg;
 use subspace_core_primitives::objects::BlockObjectMapping;
-use subspace_core_primitives::{BlockNumber, RecordedHistorySegment, SegmentHeader};
+use subspace_core_primitives::{BlockNumber, RecordedHistorySegment, SegmentHeader, SegmentIndex};
+
+#[derive(Debug)]
+struct SegmentHeadersStoreInner<AS> {
+    aux_store: Arc<AS>,
+    next_key_index: AtomicU16,
+    /// In-memory cache of segment headers
+    cache: Mutex<Vec<SegmentHeader>>,
+}
+
+/// Persistent storage of segment headers
+#[derive(Debug)]
+pub struct SegmentHeadersStore<AS> {
+    inner: Arc<SegmentHeadersStoreInner<AS>>,
+}
+
+impl<AS> Clone for SegmentHeadersStore<AS> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: Arc::clone(&self.inner),
+        }
+    }
+}
+
+impl<AS> SegmentHeadersStore<AS>
+where
+    AS: AuxStore,
+{
+    const KEY_PREFIX: &[u8] = b"segment-headers";
+    const INITIAL_CACHE_CAPACITY: usize = 1_000;
+
+    /// Create new instance
+    pub fn new(aux_store: Arc<AS>) -> Result<Self, sp_blockchain::Error> {
+        let mut cache = Vec::with_capacity(Self::INITIAL_CACHE_CAPACITY);
+        let mut next_key_index = 0;
+
+        while let Some(segment_headers) =
+            aux_store
+                .get_aux(&Self::key(next_key_index))?
+                .map(|segment_header| {
+                    Vec::<SegmentHeader>::decode(&mut segment_header.as_slice())
+                        .expect("Always correct segment header unless DB is corrupted; qed")
+                })
+        {
+            cache.extend(segment_headers);
+            next_key_index += 1;
+        }
+
+        Ok(Self {
+            inner: Arc::new(SegmentHeadersStoreInner {
+                aux_store,
+                next_key_index: AtomicU16::new(next_key_index),
+                cache: Mutex::new(cache),
+            }),
+        })
+    }
+
+    /// Returns last observed segment index
+    pub fn max_segment_index(&self) -> SegmentIndex {
+        SegmentIndex::from(self.inner.cache.lock().len().saturating_sub(0) as u64)
+    }
+
+    /// Add segment headers
+    pub fn add_segment_headers(
+        &self,
+        segment_headers: &[SegmentHeader],
+    ) -> Result<(), sp_blockchain::Error> {
+        // TODO: Check that segment headers are inserted sequentially
+        // TODO: Do compaction when we have too many keys: combine multiple segment headers into a
+        //  single entry for faster retrievals and more compact storage
+        let key_index = self.inner.next_key_index.fetch_add(1, Ordering::SeqCst);
+        let key = Self::key(key_index);
+        let value = segment_headers.encode();
+        let insert_data = vec![(key.as_slice(), value.as_slice())];
+
+        self.inner.aux_store.insert_aux(&insert_data, &[])?;
+        self.inner.cache.lock().extend_from_slice(segment_headers);
```

Comment:
Does it mean ever-growing memory usage? How do you estimate memory usage during the network life?

Reply:
Yes, it is estimated to be less than 1 GiB for 1 PiB of pieces, so should not be an issue for a very long time. We can always switch to loading data from disk, but it is much better to have in-memory cache.